system-addon/lib/ActivityStream.jsm

```diff
@@ -280,7 +280,9 @@ this.ActivityStream = class ActivityStream {
     try {
       this._updateDynamicPrefs();
       this._defaultPrefs.init();
-      this._storage.init();
+
+      // Accessing the db causes the tables to be created / migrated
+      this._storage.db; // eslint-disable-line no-unused-expressions
```

Comment:
@piatra should this `await`?

Reply:
That's what I expected too but no. The callback function passed in is [`onupgradeneeded`](https://searchfox.org/mozilla-central/rev/7e663b9fa578d425684ce2560e5fa2464f504b34/toolkit/modules/IndexedDB.jsm#245-249)
```
when the database needs to be created, or its schema needs to be upgraded.
```
So the very first call sets up the objectStores, the rest will only connect to them.

Reply:
But isn't a new `IndexedDB.open` being called for every new instance of `ActivityStreamStorage` in each of the feeds that use it? I suppose a related question is why do we need so many instances?

Reply:
1) Yes `open` is called but the cb we pass in is not. After the first call subsequent requests will only [open the db](https://searchfox.org/mozilla-central/rev/7e663b9fa578d425684ce2560e5fa2464f504b34/toolkit/modules/IndexedDB.jsm#254) requested (and expect it to be there).
2) Initially it was only needed by snippets when the code was added and the next consumer would have been the Sections. Because the SectionsManager (where the sections prefs are built) is not connected to the store I did not think to make the Storage a feed (I'm not sure if this is what you had in mind). Also this way there is also separation between the storages of different consumers.
* It's not great as I'm finding out in #4060 where to get access to the TelemetryFeed and report IndexedDB errors I have to pass the feed around to the storage.

Reply:
Oh, I guess the version is shared among all the.. tables/objects? So the earliest instance/open/request is going to trigger the upgrade, and it will need to upgrade the schema for each thing we care about?

Reply:
Yes.

To answer the initial question: also yes. I wasn't sure if there are any negative side effect of awaiting in our main `init` function. The init call happens before any of the feeds get initialized and also testing it out seemed that it could work without await-ing on storage.